rclcpp/doc/multi_threaded_executor_redesign.rst

```diff
@@ -0,0 +1,62 @@
+======================================
+:code:`MultiThreadedExecutor` Redesign
+======================================
+
+:code:`MultiThreadedExecutor` may attempt to execute one executable object multiple times. 
+This occurs because an executable object can be identified as ready in multiple threads before it has been executed, and each of these threads will then try to execute it.
+To fix this, a general solution should do the following:
+
+* Work for all the types of executable objects: timers, subscriptions, services, clients, and waitables
+* Avoid spurious wake ups (which can be costly)
+
+Summary of discussion so far
+----------------------------
+
+Broadly speaking, two approaches were discussed:
+
+1. **Handle no data in the executor**: If more than one thread is trying to execute the same object and that object is consumed after it is executed once, check for the executable before trying to execute it:
+
+    .. code-block:: c++
+
+        void execute() {
+          if(buffer->has_data()) {
+            auto data = buffer->consume_data();
+            run_any_callback(data);
+          } else {
+            // do nothing
+          }
+        }
+
+   :Advantages: 
+     * simplicity
+     * extends interface in a way that generalizes well for all other types of executable objects (i.e., timers)
+   :Disadvantages: 
+     * spurious wake up calls
+     * needs to be documented well so every class that executes :code:`AnyExecutable` objects is thread-safe and doesn't hide a race condition
+
+
+2. **Take the data**: When an executable is ready, take it and store it so that it can be executed when execute is called:
+
+    .. code-block:: c++
+
+        void take_data(std::shared_ptr<void> & data) {
+          data = buffer->consume_data();
+        }
+
+        void execute(std::shared_ptr<void> & data) {
+          run_any_callback(data_);
+        }
+
+   :Advantages:
+     * no spurious wake ups
+   :Disadvantages:
+     * more complex
+     * extends the interface in a way that doesn't generalize well for all other types of executable objects (i.e., timers)
+     * type erasure of data
+
+Ideas
+-----------------------
+
+* Add a preparation method to get the event after :code:`take_data`
+* Take a lambda that can be executed later
+* Store conditions in the waitset (like DDS)
```

Comment:
I think that having the concept of conditions (like DDS) is nice, but I'm not sure what this item means exactly.

Reply:
We spoke about it in our meeting. I'm not sure that I understand it well enough to explain. 

It was something like, we could do what DDS does, which is use conditions that can be peaked at or read. Once a condition is read, it is removed from the queue, so they don't come up again. I believe this would avoid spurious wakeups. @wjwwood, is this correct?

Reply:
Currently in a ROS 2 wait set you can add different kind of entities (subscription, service, ...).
In `DDS`, you can only [add conditions](https://community.rti.com/static/documentation/connext-dds/5.3.0/doc/api/connext_dds/api_cpp2/classdds_1_1core_1_1cond_1_1WaitSet.html#acd93c6ad7eb1da7f1f942b57e3388da6) to a WaitSet.
[Conditions API](https://community.rti.com/static/documentation/connext-dds/5.3.0/doc/api/connext_dds/api_cpp2/classdds_1_1core_1_1cond_1_1Condition.html) is quite simple, you can only check if them are ready or dispatch an attached callback.

There's no concept of "reading" or "peaking" a condition AFAIU.
I think that having something similar to DDS conditions will simplify our custom handling of different cases in wait set and executor code a lot (and the abstractions looks nice too :smiley: ).